Improve error message for partitioning by nested field in Iceberg

ebyhr opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Fixes #15091

ebyhr commented:
CI hit #14441

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -970,6 +970,14 @@ public void testCreatePartitionedTableWithNestedTypes()
+    public void testCreatePartitionedTableWithNestedField()
+    {
+        assertQueryFails(
+                "CREATE TABLE test_partitioned_table_nested_field(parent ROW(child VARCHAR)) WITH (partitioning = ARRAY['\"parent.child\"'])",
```
Just to be on the safe side, we can also add an extra assertion with a 2 level nesting - grandparent.parent.child - I tried it and works as expected.

alexjo2144 reviewed (approved):
If you don't mind adding some Spark compatibility test cases, here or in a separate PR:

- Create a table in Spark partitioned on a nested field
  - Trino should be able to read it
  - Trino should not allow insert/update/delete/merge operations on it

ebyhr commented:
Fixing `TestIcebergSparkCompatibility.testPartitionedByNestedFiled ` now.

ebyhr commented:
@alexjo2144 It seems the connector couldn't read such tables. I added another commit to fix this issue and updated more code to handle other statements. Could you take another look when you have time?

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -477,15 +479,15 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
-                    .filter(column -> partitionSourceIds.contains(column.getId()))
-                    .collect(toImmutableMap(IcebergColumnHandle::getId, identity()));
-
+        if (!partitionSourceIds.isEmpty() && !columns.isEmpty()) {
```
I find the current solution a bit hard to understand just by reading the code. It is rather unclear why the `columns` is supposed to be empty when `partitionSourceIds` is not empty. 
Maybe adding an extra comment would help?

Unrelated to the current PR: When Trino is going to be able to deal with nested fields as partitions is the output of `getColumns(...)` method supposed to change?

ebyhr replied:
> why the columns is supposed to be empty when partitionSourceIds is not empty.

Do you mean "columns is **_not_** supposed to be empty when `partitionSourceIds` is not empty"?

Yes, the method `getColumns` will be changed when we support nested fields as partitions.

findinpath replied:
> Do you mean "columns is not supposed to be empty when partitionSourceIds is not empty"?

I mean that something that would hint a code maintainer that the partition source ids may point to a nested field which will not be returned as a column handle may by `getColumns(..)` may help.

alexjo2144 replied:
I think the existing name `columns` is just not descriptive. Maybe something like `identityPartitionColumns`?

It also may be a redundant check. If `partitionSourceIds` is not empty it's impossible for `columns` to be empty. So the second block of the if here could be a `verify` instead.

ebyhr replied:
>  If partitionSourceIds is not empty it's impossible for columns to be empty.

No, that's possible in case of nested fields.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:
```diff
@@ -668,4 +669,17 @@ public static String fileName(String path)
+
+    public static void verifyNotPartitionedByNestedField(Map<Integer, Integer> indexParents, PartitionField field)
+    {
+        if (indexParents.containsKey(field.sourceId())) {
```
I'd be rather inclined to duplicate this check within `io.trino.plugin.iceberg.IcebergMetadata#getWriteLayout` then to make this method `public`. 
The `indexParents` map parameter is rather abstract.

findinpath commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -422,6 +422,43 @@ public void testTrinoReadsSparkPartitionedTable(StorageFormat storageFormat, int
+                .hasMessageContaining("Partitioning by nested field is unsupported parent.nested");
+        assertQueryFailure(() -> onTrino().executeQuery("UPDATE " + trinoTableName + " SET id = 2"))
+                .hasMessageContaining("Partitioning by nested field is unsupported parent.nested");
+        assertQueryFailure(() -> onTrino().executeQuery("DELETE FROM " + trinoTableName))
```
Out of curiosity: In a previous version of the PR before this last commit, the `DELETE` operation was still working. Why did it get removed?

ebyhr replied:
Relates to #15178. My branch status was before 84848cbafb85c4c78f1b48ebff2e4d7d289dccb1

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSinkProvider.java:
```diff
@@ -83,6 +84,7 @@ private ConnectorPageSink createPageSink(ConnectorSession session, IcebergWritab
         Schema schema = SchemaParser.fromJson(tableHandle.getSchemaAsJson());
         String partitionSpecJson = tableHandle.getPartitionsSpecsAsJson().get(tableHandle.getPartitionSpecId());
         PartitionSpec partitionSpec = PartitionSpecParser.fromJson(schema, partitionSpecJson);
+        verifyNotPartitionedByNestedField(schema, partitionSpec);
```
Sink setup feels a little late in the query execution to be catching this. Can we have these checks in IcebergMetadata, [like this one](https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L1711)?

ebyhr replied:
Removed the 1st commit.

alexjo2144 reviewed (commented):
Should we wait for a comment on if this should be supported from the Iceberg folks before merging the first commit? https://github.com/trinodb/trino/issues/15109#issuecomment-1327730105

electrum commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -1810,6 +1819,16 @@ private static void validateNotModifyingOldSnapshot(IcebergTableHandle table, Ta
+        Map<Integer, Integer> indexParents = indexParents(schema.asStruct());
+        for (PartitionField field : partitionSpec.fields()) {
+            if (indexParents.containsKey(field.sourceId())) {
+                throw new TrinoException(NOT_SUPPORTED, "Partitioning by nested field is unsupported " + field.name());
```
Nit: add a colon before the column name

```java
throw new TrinoException(NOT_SUPPORTED, "Partitioning by nested field is unsupported: " + field.name());
